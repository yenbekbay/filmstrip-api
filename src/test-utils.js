/* @flow */

import _ from 'lodash/fp';
import type from 'type-detect';

const childTypesForArray = _.memoize(_.flow(_.map(type), _.uniq));

const modelFromObject = (obj: ?Object) =>
  _.mapValues(
    _.cond([
      [
        _.overEvery([
          _.flow(type, _.eq('Array')),
          _.flow(childTypesForArray, _.head, _.eq('Object')),
        ]),
        (val: Array<any>) => [_.flow(_.head, modelFromObject)(val)],
      ],
      [
        _.overEvery([
          _.flow(type, _.eq('Array')),
          _.flow(childTypesForArray, _.size, _.eq(1)),
        ]),
        (val: Array<any>) =>
          `Array<${_.flow(childTypesForArray, _.head)(val)}>`,
      ],
      [_.flow(type, _.includes(_, ['Object', 'Array'])), modelFromObject],
      [_.stubTrue, type],
    ]),
    obj,
  );

export {modelFromObject}; // eslint-disable-line import/prefer-default-export
