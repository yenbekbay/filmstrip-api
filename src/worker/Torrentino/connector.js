/* @flow */

import cheerio from 'cheerio';
import DataLoader from 'dataloader';
import PromiseThrottle from 'promise-throttle';
import randomUseragent from 'random-useragent';
import retry from 'async-retry';
import rp from 'request-promise-native';

const TORRENTINO_ROOT = 'http://www.torrentino.me';

class HtmlConnector {
  _htmlThrottleQueue = new PromiseThrottle({
    requestsPerSecond: 0.5,
    promiseImplementation: Promise,
  });
  _htmlRp = rp.defaults({
    headers: {
      'User-Agent': randomUseragent.getRandom(),
    },
    gzip: true,
  });

  htmlLoader: { load: (url: string) => Promise<any> } = new DataLoader(
    (optionsHashes: Array<string>) => this._htmlThrottleQueue.addAll(
      optionsHashes.map((optionsHash: string) => () => retry(
        async (bail: (err: Error) => Promise<any>) => {
          try {
            const res = await this._htmlRp(JSON.parse(optionsHash));

            return cheerio.load(res);
          } catch (err) {
            if (err.statusCode === 403 || err.statusCode === 404) {
              return bail(err);
            }

            throw err;
          }
        },
        { retries: 5 },
      )),
    ), {
      batch: false,
    },
  );

  htmlGet = (
    path: string,
    query: void | { [key: string]: mixed },
  ) => this.htmlLoader.load(
    JSON.stringify({
      uri: `${TORRENTINO_ROOT}/${path}`,
      qs: query || {},
    }),
  );
}

export default HtmlConnector;
