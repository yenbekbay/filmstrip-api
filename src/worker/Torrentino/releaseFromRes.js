/* @flow */

import _ from 'lodash/fp';
import bytes from 'bytes';

import { torrentQualityTest } from '../utils';
import type { TorrentinoRelease } from '../../types';

const separatorRegex = /\s*,\s*|\s+/;
const validAudioTranslationTypes = [
  'Лицензия',
  'Дублированный',
  'Профессиональный многоголосый',
];

const movieInfoFromRes = ($: () => Object) => {
  const headNode = $('.main > section > .entity > .head-plate > .head');
  const detailsNode = headNode.find('.specialty');

  const title = headNode.find('.header-group > h1').text() || null;
  const year = parseInt(
    detailsNode.find('tr.clause td:contains("Год") ~ td').text(), 10,
  );

  if (!title || !year) return null;

  const posterUrl = headNode.find('.cover > img').attr('src') || null;
  const originalTitle = headNode.find('.header-group > h2').text() || null;
  const synopsis = detailsNode.children('p').text() || null;
  const [hours, minutes] = _.split(
    ':',
    detailsNode.find('tr.clause td:contains("Длительность") ~ td').text(),
  );
  const runtime = (parseInt(hours, 10) * 60) + (parseInt(minutes, 10) * 60);
  const productionCountries = _.flow(
    _.split(','),
    _.map(_.flow(
      _.trim,
      _.capitalize,
      _.replace('Сша', 'США'),
    )),
  )(detailsNode.find('tr.clause td:contains("Страна") ~ td').text());
  const genres = _.flow(
    _.split(','),
    _.map(_.trim),
  )(detailsNode.find('tr.clause td:contains("Жанр") ~ td').text());
  const releaseDate = _.flow(
    _.split('/'),
    ([dateDay, dateMonth, dateYear]: Array<?string>) => (
      (dateDay && dateMonth && dateYear)
        ? [dateYear, dateMonth, dateDay].join('-')
        : null
    ),
  )(detailsNode.find('tr.clause td:contains("Премьера в РФ") ~ td').text());

  const creditsForJob = (job: string) => _.flow(
    _.split(','),
    _.map(_.flow(
      _.trim,
      (name: ?string) => (
        name ? { name, photoUrl: null } : null
      ),
    )),
    _.compact,
  )(detailsNode.find(`tr.clause td:contains("${job}") ~ td`).text());

  const directors = creditsForJob('Режиссер');
  const cast = creditsForJob('В ролях');

  return {
    title,
    originalTitle,
    posterUrl,
    synopsis,
    year,
    runtime,
    productionCountries,
    genres,
    releaseDate,
    credits: {
      cast,
      crew: {
        directors,
      },
    },
  };
};

const torrentsFromRes = ($: () => Object) =>
  $('.main > section > .entity > .list-start > .table-list')
    .first()
    .find('tr.item')
    .get()
    .map((el: Object) => {
      const quality = _.flow(
        _.split('x'),
        _.head,
        (videoWidth: ?string) => (videoWidth ? parseInt(videoWidth, 10) : null),
        _.cond([
          [_.gt(_, 1900), _.constant('1080p')],
          [_.gt(_, 1000), _.constant('720p')],
          [_.stubTrue, _.constant('SD')],
        ]),
      )($(el).children('.video').text() || null);

      const audioTranslationType = _.flow(
        (audio: ?string) => (audio === '.  .' ? null : audio),
      )($(el).children('.audio').text() || null);
      const audioTracks = _.flow(
        _.split(separatorRegex),
        _.map((language: string) => (language === 'null' ? null : language)),
        _.compact,
      )($(el).children('.languages').text());
      const bundledSubtitles = _.flow(
        _.split(separatorRegex),
        _.compact,
      )($(el).children('.subtitles').text());
      const size = _.flow(
        (sizeInGb: string) => (
          _.includes('ГБ', sizeInGb)
            ? bytes(sizeInGb.replace(/\s+/, '').replace('ГБ', 'GB'))
            : NaN
        ),
      )($(el).children('.size').text());
      const seeds = parseInt($(el).find('.seed-leech > .seed').text(), 10);
      const peers = parseInt($(el).find('.seed-leech > .leech').text(), 10);
      const magnetLink = $(el).find('.download > a').attr('data-default');

      return (
        quality !== 'SD' &&
        audioTranslationType &&
        _.includes('ru', audioTracks) &&
        _.includes(audioTranslationType, validAudioTranslationTypes) &&
        !!size && seeds > 0 && peers > 0 && _.startsWith('magnet:?', magnetLink)
      ) ? {
        source: 'Torrentino',
        quality,
        audioTranslationType,
        audioTracks,
        bundledSubtitles,
        size,
        seeds,
        peers,
        magnetLink,
      } : null;
    });

const releaseFromRes = (
  $: () => Object,
  torrentinoSlug: string,
): ?TorrentinoRelease => {
  const movieInfo = movieInfoFromRes($);
  if (!movieInfo) return null;

  const torrents = torrentsFromRes($) || [];

  return {
    info: {
      ...movieInfo,
      kpId: parseInt(_.head(torrentinoSlug.match(/\d+/)), 10),
      torrentinoSlug,
    },
    torrents: _.flow(
      _.compact,
      _.filter(torrentQualityTest),
      _.orderBy(['seeds'], ['desc']),
    )(torrents),
  };
};

export default releaseFromRes;
