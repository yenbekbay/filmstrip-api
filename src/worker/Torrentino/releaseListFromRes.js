/* @flow */

import _ from 'lodash/fp';

const releaseListFromRes = ($: () => Object) => {
  const movieNodes = $(
    '.main > section > .showcase .tiles > .tile[data-movie-id]',
  ).get();

  return _.compact(
    movieNodes.map((el: Object) => {
      const torrentinoSlug = _.replace(
        '/movie/', '', $(el).children('a').first().attr('href'),
      );
      const title = $(el).find('a > .title > .name').text();
      const year = parseInt($(el).find('a > .title > .year').text(), 10);
      const kpId = parseInt(_.head(torrentinoSlug.match(/\d+/)), 10);

      if (!torrentinoSlug || !title || !year || !kpId) return null;

      return { torrentinoSlug, kpId, title, year };
    }),
  );
};

export default releaseListFromRes;
