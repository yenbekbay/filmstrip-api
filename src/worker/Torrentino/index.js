/* @flow */

import _ from 'lodash/fp';
import retry from 'async-retry';

import releaseFromRes from './releaseFromRes';
import releaseListFromRes from './releaseListFromRes';
import TorrentinoConnector from './connector';

class Torrentino {
  _connector = new TorrentinoConnector();

  getLatestReleases = async (
    page: void | number,
  ): Promise<Array<Object>> => retry(
    async (bail: (err: Error) => void) => {
      const currentYear = new Date().getFullYear();
      const noReleasesErrMessage = 'Failed to get latest Torrentino releases';

      try {
        const $ = await this._connector.htmlGet(
          'movies',
          {
            quality: 'hq',
            years: `${currentYear - 1},${currentYear}`,
            sort: 'date',
            page: page || 1,
          },
        );

        const releases = releaseListFromRes($);

        if (releases.length === 0) {
          throw new Error(noReleasesErrMessage);
        }

        return releases;
      } catch (err) {
        if (err.message === noReleasesErrMessage) {
          throw err;
        }

        return bail(err);
      }
    },
    { retries: 5 },
  );

  getReleaseDetails = async (slug: string) => retry(
    async (bail: (err: Error) => void) => {
      const noReleaseErrMessage =
        `Failed to get Torrentino release for movie ${slug}`;

      try {
        const $ = await this._connector.htmlGet(`movie/${slug}`);

        const release = releaseFromRes($, slug);
        if (!release) {
          throw new Error(noReleaseErrMessage);
        }

        return release;
      } catch (err) {
        if (err.message === noReleaseErrMessage) {
          throw err;
        }

        return bail(err);
      }
    },
    { retries: 5 },
  );

  getTorrentinoSlug = async (
    query: { title: string, kpId: number },
  ): Promise<?string> => {
    const $ = await this._connector.htmlGet('search', {
      type: 'movies',
      search: query.title,
    });

    const releases = releaseListFromRes($).filter(
      ({ kpId }: Object) => kpId === query.kpId,
    );

    return _.flow(_.head, _.get('torrentinoSlug'))(releases);
  };
}

export default Torrentino;
