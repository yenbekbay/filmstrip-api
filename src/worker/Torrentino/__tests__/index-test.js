/* @flow */

import {modelFromObject} from '../../../test-utils';
import Torrentino from '../';

const sampleTorrentinoSlug = '854942-jason-bourne';
const sampleSearchQuery = {
  title: 'Бэтмен против Супермена: На заре справедливости',
  kpId: 770631,
};

describe('Torrentino', () => {
  let torrentino: Torrentino;

  beforeAll(() => {
    torrentino = new Torrentino();
  });

  it('fetches a list of latest releases', async () => {
    const torrentinoReleases = await torrentino.getLatestReleases();
    expect(modelFromObject({torrentinoReleases})).toMatchSnapshot();
  });

  it('fetches release details for a given torrentino slug', async () => {
    const release = await torrentino.getReleaseDetails(sampleTorrentinoSlug);
    expect(modelFromObject(release)).toMatchSnapshot();
  });

  it('finds a releases by a given query', async () => {
    const torrentinoSlug = await torrentino.getTorrentinoSlug(
      sampleSearchQuery,
    );
    expect(torrentinoSlug).toMatchSnapshot();
  });
});
