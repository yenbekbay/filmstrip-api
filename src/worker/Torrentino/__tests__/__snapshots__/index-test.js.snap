exports[`Torrentino fetches a list of latest releases 1`] = `
Object {
  "torrentinoReleases": Array [
    Object {
      "kpId": "number",
      "title": "string",
      "torrentinoSlug": "string",
      "year": "number",
    },
  ],
}
`;

exports[`Torrentino fetches release details for a given torrentino slug 1`] = `
Object {
  "info": Object {
    "credits": Object {
      "cast": Array [
        Object {
          "name": "string",
          "photoUrl": "null",
        },
      ],
      "crew": Object {
        "directors": Array [
          Object {
            "name": "string",
            "photoUrl": "null",
          },
        ],
      },
    },
    "genres": "Array<string>",
    "kpId": "number",
    "originalTitle": "string",
    "posterUrl": "string",
    "productionCountries": "Array<string>",
    "releaseDate": "string",
    "runtime": "number",
    "synopsis": "string",
    "title": "string",
    "torrentinoSlug": "string",
    "year": "number",
  },
  "torrents": Array [
    Object {
      "audioTracks": "Array<string>",
      "audioTranslationType": "string",
      "bundledSubtitles": "Array<string>",
      "magnetLink": "string",
      "peers": "number",
      "quality": "string",
      "seeds": "number",
      "size": "number",
      "source": "string",
    },
  ],
}
`;

exports[`Torrentino finds a releases by a given query 1`] = `"770631-batman-v-superman-dawn-justice"`;
